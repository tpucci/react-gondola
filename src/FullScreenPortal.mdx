---
name: FullScreenPortal
menu: API
route: /api/full-screen-portal
---

# FullScreenPortal

`type: React.Component`

## Props

| Prop name  | Required | Default     | Usage                                                                                                      |
| :--------: | :------: | ----------- | ---------------------------------------------------------------------------------------------------------- |
| `children` |  `true`  | `undefined` | Children are wrapped in a `React.Fragment`. All full screen Stops are rendered in front of those children. |

## Properties

### `fullScreenStack: IStreamListener<StopInterface[]>`

MobX Strem Listener of list of React Gondola full screen Stops. This stream is observed by MobX and fires the rendering of the full screen stack in the `FullScreenPortal.render` method.

## Basic usage

Wrap your application main canal with `FullScreenPortal`. `FullScreenPortal` must appear at the top of your tree in order to render full screen pages.

```tsx
import React, { Component } from 'react';
import { FullScreenPortal } from 'react-nonav';

import { MainCanal } from './canals/MainCanal';

export default class App extends Component {
  render() {
    return (
      <FullScreenPortal>
        <MainCanal />
      </FullScreenPortal>
    );
  }
}
```
