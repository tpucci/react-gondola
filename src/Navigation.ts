import { Subject, Observable } from 'rxjs';
import { ComponentType } from 'react';
import { mergeAll, map, scan } from 'rxjs/operators';

export interface IStop {
  name: string;
  Component: ComponentType;
  isFullScreen?: boolean;
}

type Stack = IStop[];

interface IFullScreenStackProperties {
  canalId: string;
  fullScreenStack: Stack;
}

interface IFullScreenStackMap {
  [canalId: string]: Stack;
}

export class Navigation {
  static getInstance() {
    if (Navigation.instance) {
      return Navigation.instance;
    }
    Navigation.instance = new Navigation();
    return Navigation.instance;
  }
  private static instance: Navigation;

  canalsFullScreenStackProperties$ = new Subject<
    Observable<IFullScreenStackProperties>
  >();

  fullSceenStack$: Observable<
    Stack
  > = this.canalsFullScreenStackProperties$.pipe(
    mergeAll(),
    scan(
      (
        fullScreenStackMap: IFullScreenStackMap,
        fullScreenStackProperties: IFullScreenStackProperties
      ) => ({
        ...fullScreenStackMap,
        [fullScreenStackProperties.canalId]:
          fullScreenStackProperties.fullScreenStack
      }),
      {}
    ),
    map(fullScreenStackMap =>
      Object.keys(fullScreenStackMap).reduce(
        (fullScreenStack: Stack, canalId: string) =>
          fullScreenStack.concat(fullScreenStackMap[canalId]),
        []
      )
    )
  );
}
