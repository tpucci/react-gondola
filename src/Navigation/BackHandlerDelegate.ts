import { BackHandler } from 'react-native';
import { fromEventPattern, Observable, Subject } from 'rxjs';
import { share, map, merge, tap } from 'rxjs/operators';

export interface BackEvent {
  target: string | null;
}

export class BackHandlerDelegate {
  /**
   * Attach the given handler to the event source. When the source emits,
   * calls the handler. The handler is responsible to register a callback.
   * If it succeeded to register a callback, we call it.
   * If not, and no other react native back event listener has been registered
   * elsewhere, this listener returns false which might kill the app.
   * See https://facebook.github.io/react-native/docs/backhandler.html.
   * @param handler
   */
  private addEventListener = (handler: () => void) => {
    BackHandler.addEventListener('hardwareBackPress', () => {
      handler();
      return true;
    });
  };

  private softwareBack$ = new Subject<undefined>();

  private back$: Observable<BackEvent> = fromEventPattern<undefined>(this.addEventListener).pipe(
    merge(this.softwareBack$),
    tap(() => {
      // After the event has spread, we fire onBacks callbacks.
      setTimeout(() => {
        if (this.onBackCallback) {
          this.onBackCallback();
          this.onBackCallback = undefined;
        } else {
          BackHandler.exitApp();
        }
      }, 0);
    }),
    map(() => ({
      target: null,
    })),
    share()
  );

  private onBackCallback?: () => any;
  setOnBackCallback = (cb: () => any) => {
    this.onBackCallback = cb;
  };

  back = () => {
    this.softwareBack$.next();
  };

  defaultBackContext = { back$: this.back$ };
}
