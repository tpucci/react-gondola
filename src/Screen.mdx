---
name: Screen
menu: API
route: /api/screen
---

import { Props } from 'docz';
import { Screen } from 'react-nonav';

# Screen

`type: React.Component`

## Props

|    Prop name    | Required | Default     | Usage                                                                                                                                                                                     |
| :-------------: | :------: | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `isFullScreen`  | `false`  | `false`     | Specifies if the screen should be render in the `FullScreenPortal`.                                                                                                                       |
|    `onBack`     | `false`  | `undefined` | Function to call when a back event occurs. Use this prop and `visible` to control your Canal.                                                                                             |
|     `name`      |  `true`  | `undefined` | Unique identifier of the screen.                                                                                                                                                          |
|   `Component`   |  `true`  | `undefined` | Component to render. This is your page.                                                                                                                                                   |
|     `props`     | `false`  | `undefined` | Props to pass to the Component. Those props are transition-safe. This means while a transition occurs, those props are frozen. When the transition completes, the props flow is released. |
| `Transtitioner` | `false`  | `None`      | Transition to apply to the screen when it appears or disappears.                                                                                                                          |
|    `visible`    |  `true`  | `false`     | Visibility of your screen. Use this prop and `onBack` to control your Canal.                                                                                                              |
