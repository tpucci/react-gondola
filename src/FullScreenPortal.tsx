import React, { Component as ReactComponent, Fragment } from 'react';
import { View, StyleSheet } from 'react-native';
import { Observer } from 'mobx-react/native';
import { fromStream } from 'mobx-utils';
import { Navigation } from './Navigation';

export class FullScreenPortal extends ReactComponent {
  fullScreenStack = fromStream(Navigation.getInstance().fullSceenStack$);

  render() {
    return (
      <View style={StyleSheet.absoluteFill}>
        <Fragment>{this.props.children}</Fragment>
        <Observer>
          {() => {
            if (this.fullScreenStack.current) {
              return (
                this.fullScreenStack.current &&
                this.fullScreenStack.current.map(({ Component, name }) => (
                  <View style={StyleSheet.absoluteFill} key={name}>
                    <Component />
                  </View>
                ))
              );
            }
            return null;
          }}
        </Observer>
      </View>
    );
  }
}
