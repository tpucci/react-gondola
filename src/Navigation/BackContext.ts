import { createContext } from 'react';
import { Observable } from 'rxjs';
import { IBackEvent } from './BackHandlerDelegate';
import { Navigation } from './Navigation';

interface IBackContext {
  back$: Observable<IBackEvent>;
}

export const BackContext = createContext<IBackContext>(
  Navigation.instance.backHandlerDelegate.defaultBackContext
);
