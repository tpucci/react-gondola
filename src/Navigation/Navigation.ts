import { FullScreenDelegate } from './FullScreenDelegate';
import { BackHandlerDelegate } from './BackHandlerDelegate';

export class Navigation {
  static get instance() {
    if (Navigation.singleton) {
      return Navigation.singleton;
    }
    Navigation.singleton = new Navigation();
    return Navigation.singleton;
  }
  private static singleton: Navigation;

  fullScreenDelegate = new FullScreenDelegate();

  backHandlerDelegate = new BackHandlerDelegate();

  private constructor() {}
}
