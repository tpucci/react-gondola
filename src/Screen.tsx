import React, { Component as ReactComponent, createFactory, ComponentType } from 'react';
import { BackContext } from './Navigation/BackContext';
import { filter, share, tap } from 'rxjs/operators';
import { Navigation } from './Navigation';
import { TransitionComponentType } from './transitions/Transition';
import { None } from './transitions';
import { withBackContext, WithBackContext } from './withBackContext';

export interface ScreenProps {
  isFullScreen?: boolean;
  onBack?: (() => any) | undefined;
  name: string;
  Component: ComponentType<any>;
  props?: object | undefined;
  Transitioner?: TransitionComponentType;
  visible: boolean;
}

class ScreenComponent extends ReactComponent<WithBackContext<ScreenProps>> {
  static defaultProps = {
    isFullScreen: false,
    Transitioner: None,
    props: {},
  };

  back$ = this.props.backContext.back$.pipe(
    filter(value => value.target === this.props.name),
    tap(() => {
      if (this.props.onBack) {
        Navigation.instance.backHandlerDelegate.setOnBackCallback(this.props.onBack);
      }
    }),
    share()
  );

  backSubscription = this.back$.subscribe();

  componentWillUnmount() {
    this.backSubscription.unsubscribe();
  }

  /**
   * @TODO 2019-07-26 Update @types/react once https://github.com/DefinitelyTyped/DefinitelyTyped/pull is merged.
   */
  // @ts-ignore
  factory = createFactory(this.props.Component);

  render() {
    const { Transitioner, onBack, props } = this.props;
    return (
      /**
       *  @todo Transitioner is always defined in static defaultProps.
       */
      // @ts-ignore
      <Transitioner directionForward={this.props.visible}>
        <BackContext.Provider
          value={{
            back$: this.back$,
          }}
        >
          {this.factory({ navigation: { goBack: onBack }, ...props })}
        </BackContext.Provider>
      </Transitioner>
    );
  }
}

export const Screen = withBackContext(ScreenComponent);
