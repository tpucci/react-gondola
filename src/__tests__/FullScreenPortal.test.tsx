import React from 'react';
import TestRenderer from 'react-test-renderer';

/**
 * import order is important between recompose and rxjs
 * @see https://stackoverflow.com/questions/53878650/you-provided-an-invalid-object-where-a-stream-was-expected-when-using-rxjs6-an.
 */
import 'recompose';
import { Subject } from 'rxjs';
import { View, Text } from 'react-native';

import { FullScreenPortal } from '../FullScreenPortal';
import { Navigation } from '../Navigation';
import { BackEvent } from '../Navigation/BackHandlerDelegate';
import { BackContext } from '../Navigation/BackContext';
import { Canal } from '../Canal';
import { Screen } from '../Screen';

describe('FullScreenPortal', () => {
  it('renders the full screen portal', () => {
    const testRenderer = TestRenderer.create(
      <FullScreenPortal>
        <View></View>
      </FullScreenPortal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('renders the full screen screens', () => {
    const testRenderer = TestRenderer.create(
      <FullScreenPortal>
        <View></View>
      </FullScreenPortal>
    );
    Navigation.instance.fullScreenDelegate.canalsFullScreenStackProperties$.next({
      canalId: '1',
      fullScreenStack: [
        <View key="a">
          <Text>I should be rendered</Text>
        </View>,
      ],
    });
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('passes on back events', () => {
    const back$ = new Subject<BackEvent>();
    const spy = jest.fn();
    const testRenderer = TestRenderer.create(
      <BackContext.Provider value={{ back$ }}>
        <FullScreenPortal>
          <Canal>
            <Screen Component={() => <Text>a</Text>} name="a" visible isFullScreen />
          </Canal>
        </FullScreenPortal>
      </BackContext.Provider>
    );
    // @ts-ignore
    testRenderer.root.children[0].instance.back$.subscribe(spy);
    back$.next({ target: null });
    expect(spy).toHaveBeenCalledWith({
      target: 'a',
    });
    testRenderer.update(
      <BackContext.Provider value={{ back$ }}>
        <FullScreenPortal>
          <Canal>
            <Screen Component={() => <Text>a</Text>} name="a" visible />
          </Canal>
        </FullScreenPortal>
      </BackContext.Provider>
    );
    // @ts-ignore
    testRenderer.root.children[0].instance.back$.subscribe(spy);
    back$.next({ target: null });
    expect(spy).toHaveBeenCalledWith({
      target: null,
    });
  });
});
