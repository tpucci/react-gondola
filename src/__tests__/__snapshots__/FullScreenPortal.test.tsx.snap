// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreenPortal renders the full screen portal 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
/>
`;
